src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/ICustomMarshalerTests.cs

```diff
@@ -0,0 +1,537 @@
+        public static extern int MarshalerOnBoxedValueTypeMethod([MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(BoxedValueTypeCustomMarshaler))] object i);
+
+
+        public class StringContainer
+        {
+            public string Value { get; set; }
+        }
+
+        public class ClassForwardingCustomMarshaler : ICustomMarshaler
+        {
+            public void CleanUpManagedData(object ManagedObj) {}
+            public void CleanUpNativeData(IntPtr pNativeData)
+            {
+                Marshal.ZeroFreeGlobalAllocAnsi(pNativeData);
+            }
+
+            public int GetNativeDataSize() => -1;
+
+            public IntPtr MarshalManagedToNative(object ManagedObj)
+            {
+                return Marshal.StringToCoTaskMemAnsi(((StringContainer)ManagedObj).Value);
+            }
+
+            public object MarshalNativeToManaged(IntPtr pNativeData)
+            {
+                int doubleValue = pNativeData.ToInt32() * 2;
+                return new StringContainer { Value = doubleValue.ToString() };
+            }
+
+            public static ICustomMarshaler GetInstance(string cookie) => new ClassForwardingCustomMarshaler();
+        }
+
+        [DllImport(LibcLibrary, EntryPoint = "atoi", CallingConvention = CallingConvention.Cdecl)]
+        [return: MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(ClassForwardingCustomMarshaler))]
+        public static extern StringContainer MarshalerOnClassTypeMethod([MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(ClassForwardingCustomMarshaler))] StringContainer str);
+
+        public class OrderTrackingCustomMarshaler : ICustomMarshaler
+        {
+            public static List<string> Events { get; } = new List<string>();
+            public static IntPtr MarshaledNativeData { get; set; }
+
+            public void CleanUpManagedData(object ManagedObj)
+            {
+                Events.Add("Called CleanUpManagedData");
+            }
+
+            public void CleanUpNativeData(IntPtr pNativeData)
+            {
+                Assert.Equal(MarshaledNativeData, pNativeData);
+                Marshal.ZeroFreeGlobalAllocAnsi(pNativeData);
+
+                Events.Add("Called CleanUpNativeData");
+            }
+
+            public int GetNativeDataSize()
+            {
+                Events.Add("Called GetNativeDataSize");
+                return 0;
+            }
+
+            public IntPtr MarshalManagedToNative(object ManagedObj)
+            {
+                Events.Add("Called MarshalManagedToNative");
+                MarshaledNativeData = Marshal.StringToCoTaskMemAnsi((string)ManagedObj);
+                return MarshaledNativeData;
+            }
+
+            public object MarshalNativeToManaged(IntPtr pNativeData)
+            {
+                Events.Add("Called MarshalNativeToManaged");
+                return pNativeData.ToInt32().ToString();
```

Comment:
Agreed. @hughbe Do you want to open a new PR in CoreCLR repo instead so that you could write proper native functions that do the right thing? We have interop tests in CoreCLR repo [here](https://github.com/dotnet/coreclr/tree/master/tests/src/Interop) if you are looking for examples.